### PR TITLE
Add FXIOS-5790 [v112] rust sync manager added

### PR DESCRIPTION
**Pull request by lougeniaC64**
This is the third task in #13278 which is adding the `RustSyncManager` class for the sync manager integration. This class will replace the existing `BrowserSyncManager` class.

**Note:** The changes in `RustRemoteTabs.swift`, `RustSyncManagerAPI.swift`, and `SyncTelemetryUtils.swift` should have been included in #13330 but were omitted because of the Glean build issues.

### CODE REVIEW

**Comment by lougeniaC64** [comment]
@nbhasin2 @mhammond @tarikeshaq Ready for review, and I'm open to any all feedback. This is the bulk of the sync manager integration logic, and the replacement for the current [BrowserSyncManager](https://github.com/mozilla-mobile/firefox-ios/blob/main/Providers/Profile.swift#L751) class, so this should be logic that both of our teams can easily grok and feel comfortable maintaining.

File: Providers/RustSyncManager.swift [hunk 1]
```diff
@@ -0,0 +1,635 @@
...
+import Common
+
+// Extends NSObject so we can use timers.
+public class RustSyncManager: NSObject, SyncManager {
+    // We shouldn't live beyond our containing BrowserProfile, either in the main app
+    // or in an extension.
+    // But it's possible that we'll finish a side-effect sync after we've ditched the
+    // profile as a whole, so we hold on to our Prefs, potentially for a little while
+    // longer. This is safe as a strong reference, because there's no cycle.
+    unowned fileprivate let profile: BrowserProfile
```
Comment (nbhasin2) [inline Providers/RustSyncManager.swift:20]:
```suggestion
    private let profile: BrowserProfile
```

Is there a reason you need `unowned` here? We use that mostly in closures inside the capture list `[unowned let...] in` but I am not sure what benefit we get for using unowned here. It might end up being dangerous because its a strong reference which has no relation to retain and release cycle. So you can see random crashes.

File: Providers/RustSyncManager.swift [hunk 2]
```diff
@@ -0,0 +1,635 @@
...
+    fileprivate let prefs: Prefs
+    fileprivate var constellationStateUpdate: Any?
+    fileprivate var syncTimer: Timer?
+    fileprivate var backgrounded: Bool = true
```
Comment (nbhasin2) [inline Providers/RustSyncManager.swift:24]:
```suggestion
    private let prefs: Prefs
    private var constellationStateUpdate: Any?
    private var syncTimer: Timer?
    private var backgrounded: Bool = true
```
nit ^

File: Providers/RustSyncManager.swift [hunk 3]
```diff
@@ -0,0 +1,635 @@
...
+    private let logger: Logger
+
+    let FifteenMinutes = TimeInterval(60 * 15)
```
Comment (nbhasin2) [inline Providers/RustSyncManager.swift:27]: nit: 
```suggestion
    let fifteenMinutesInterval = TimeInterval(60 * 15)
```

File: Providers/RustSyncManager.swift [hunk 4]
```diff
@@ -0,0 +1,635 @@
...
+
+    deinit {
+        if let c = constellationStateUpdate {
+            NotificationCenter.default.removeObserver(c)
+        }
```
Comment (nbhasin2) [inline Providers/RustSyncManager.swift:32]: nit: avoid single letter var
```suggestion
        if let stateUpdate = constellationStateUpdate {
            NotificationCenter.default.removeObserver(stateUpdate)
        }
```

File: Providers/RustSyncManager.swift (hunk 1 above)
Comment (lougeniaC64) [reply Providers/RustSyncManager.swift:20]: No other reason that it was in [BrowserSyncManager](https://github.com/mozilla-mobile/firefox-ios/blob/7c90d922f0b2cdad3d311f6649ecc2c04d4e20f6/Providers/Profile.swift#L757).

Comment (nbhasin2) [reply Providers/RustSyncManager.swift:20]: The interesting fact is that its an 8-year-old code that still works!

Personally, I avoid the `unowned` unless I am very confident in the lifecycle of the variable which might have been the case 8 years ago.

**Review by tarikeshaq — CHANGES_REQUESTED** [review]
Had a very brief look, will try to provide more insightful feedback early next week!

This is coming together 🎉

File: Providers/RustSyncManager.swift [hunk 5]
```diff
@@ -0,0 +1,628 @@
...
+        let center = NotificationCenter.default
+
+        center.addObserver(self,
+                           selector: #selector(onStartSyncing),
+                           name: .ProfileDidStartSyncing,
+                           object: nil)
+        center.addObserver(self,
+                           selector: #selector(onFinishSyncing),
+                           name: .ProfileDidFinishSyncing,
+                           object: nil)
```
Comment (lmarceau) [inline Providers/RustSyncManager.swift:243]: Two questions related to notifications:
1. So, we send those `.ProfileDidStartSyncing` & `.ProfileDidStartSyncing` from this `RustSyncManager` class and currently from BrowserProfile code too. Will we keep sending those notifications from BrowserProfile in the near future? 
2. If we send those notifications from `RustSyncManager`, is there added benefits of listening to the notifications `RustSyncManager`? Could we directly call `onStartSyncing` and `onFinishSyncing` instead?

File: Providers/RustSyncManager.swift [hunk 6]
```diff
@@ -0,0 +1,628 @@
...
+    }
+
+    private func getEngineEnablementChangesForAccount() -> [String: Bool] {
+        var engineEnablements: [String: Bool] = [:]
+        // We just created the account, the user went through the Choose What to Sync
+        // screen on FxA.
+        if let declined = UserDefaults.standard.stringArray(
+            forKey: "fxa.cwts.declinedSyncEngines") {
+            declined.forEach { engineEnablements[$0] = false }
+            UserDefaults.standard.removeObject(forKey: "fxa.cwts.declinedSyncEngines")
```
Comment (lmarceau) [inline Providers/RustSyncManager.swift:311]: nit, Looks like this key is used in other places, we could maybe have it as a constant?

File: Providers/RustSyncManager.swift [hunk 7]
```diff
@@ -0,0 +1,628 @@
...
+                    self.prefsForSync.setObject(nil, forKey: stateChangedPref)
+                }
+            }
+        }
+
+        if !engineEnablements.isEmpty {
+            logger.log("""
+                engines to enable:
+                \(engineEnablements.compactMap { $0.value ? $0.key : nil })
+                """,
```
Comment (lmarceau) [inline Providers/RustSyncManager.swift:329]: Wondering if this string format will save the log on multiple lines (I've had that issue in the past for other loggers), can we double check to make sure it's saved on one line?

File: Providers/RustSyncManager.swift [hunk 8]
```diff
@@ -0,0 +1,628 @@
...
+        public let description = "Failed to get token server endpoint url."
+    }
+
+    public class EngineAndKeyRetrievalError: MaybeErrorType {
+        public let description = "Failed to get sync engine and key data."
+    }
+
+    fileprivate func getEnginesAndKeys(
+        engines: [String]
+    ) -> Deferred<Maybe<([EngineIdentifier], [String: String])>> {
```
Comment (lmarceau) [inline Providers/RustSyncManager.swift:364]: Looking at the code, I guess it would be too much work to remove Deferred usage? 🥲

File: Providers/RustSyncManager.swift [hunk 9]
```diff
@@ -0,0 +1,628 @@
...
+                // Clear prefs after we're done clearing everything else -- just in case
+                // one of them needs the prefs and we race. Clear regardless of success
+                // or failure.
+
+                // This will remove keys from the Keychain if they exist, as well
+                // as wiping the Sync prefs.
+
+                // XXX: `Scratchpad.clearFromPrefs` and `clearAll` were pulled from
+                // `SyncStateMachine.clearStateFromPrefs` to reduce RustSyncManager's
+                // dependence on the swift sync state machine logid. This will make
```
Comment (lmarceau) [inline Providers/RustSyncManager.swift:290]: nit
```suggestion
                // dependence on the swift sync state machine logic. This will make
```

File: Providers/RustSyncManager.swift [hunk 10]
```diff
@@ -0,0 +1,628 @@
...
+                                    """,
+                                    level: .info,
+                                    category: .sync)
+                                self.logger.log("""
+                                    Returned telemetry:
+                                    \(String(describing: syncResult.telemetryJson ??
+                                             "(No telemetry data was returned)"))
+                                    """,
+                                    level: .info,
+                                    category: .sync)
```
Comment (lmarceau) [inline Providers/RustSyncManager.swift:487]: I think this can be done in one log only using the `extra` parameters on logger. We send info log as breadcrumbs to Sentry, so would be better to have this info noted in the same breadcrumb I think.

File: Providers/RustSyncManager.swift [hunk 11]
```diff
@@ -0,0 +1,628 @@
...
+
+    public func syncClients() -> OldSyncResult {
+        // XXX: This function exists to to comply with the `SyncManager` protocol and has
+        // no callers. It will be removed when the rust sync manager rollout is complete.
+        // To be safe, `syncClientsThenTabs` is called.
+        return self.syncClientsThenTabs()
+    }
+
+    public func syncHistory() -> OldSyncResult {
+        // XXX: The retrurn type of this function has been changed to comply with the
```
Comment (lmarceau) [inline Providers/RustSyncManager.swift:614]: nit
```suggestion
        // XXX: The return type of this function has been changed to comply with the
```

**Review by lmarceau — COMMENTED** [review]
Hello! I wanted to get familiar with the code too, I have some minor questions and nits. Let me know if there's anything. Thank you 🙏

File: Providers/RustSyncManager.swift [hunk 12]
```diff
@@ -0,0 +1,628 @@
...
+        return deferred
+    }
+
+    fileprivate func syncRustEngines(
+        why: MozillaAppServices.SyncReason,
+        engines: [String]
+    ) -> Deferred<Maybe<MozillaAppServices.SyncResult>> {
+        let deferred = Deferred<Maybe<MozillaAppServices.SyncResult>>()
+
+        syncQueue.async { [unowned self] in
```
[the remark this answers is not in the document]
Comment (lougeniaC64) [reply Providers/RustSyncManager.swift:412]: Sorry, I meant to leave I comment here. I purposely left this dispatch queue usage to illustrate an alternative because it was something Nish and I discussed and I was wondering if seeing the code might help. For sure we should only be using dispatch queues *_either_* here or in the lower level `RustSyncManagerAPI` functions.

@nbhasin2 Which is better?

File: Providers/RustSyncManager.swift [hunk 13]
```diff
@@ -0,0 +1,628 @@
...
+        case .tv:
+            return SyncManagerDeviceType.tv
+        case .unknown:
+            return SyncManagerDeviceType.unknown
+        }
+    }
+
+    @discardableResult public func syncEverything(why: OldSyncReason) -> Success {
+        if let accountManager = RustFirefoxAccounts.shared.accountManager.peek(),
+           accountManager.accountMigrationInFlight() {
```
[the remark this answers is not in the document]
Comment (lougeniaC64) [reply Providers/RustSyncManager.swift:544]: 🫠 I'll remove that--thanks!

File: Providers/RustSyncManager.swift (hunk 8 above)
Comment (lougeniaC64) [reply Providers/RustSyncManager.swift:364]: Yeah, because of the `SyncManager` protocol removing Deferred now would require a larger refactor. However, once the Nimbus sync manager integration rollout is over we can update the Deferred usage and refactor the callers.

File: Providers/RustSyncManager.swift [hunk 14]
```diff
@@ -0,0 +1,628 @@
...
+        // persisted state string. And if we do not call the sync manager's rust
+        // function with persisted state data that call will be treated as the first
+        // sync for the account, which could take longer than the user expects to be
+        // completed. To avoid that, we sync one component with the rust sync manager
+        // and store the returned persisted state data for the first sync manager full
+        // sync.
+
+        var stateHasBeenSet = false
+
+        // XXX: `RustTogglableEngines` is not used here because we want to sync the
```
[the remark this answers is not in the document]
Comment (lougeniaC64) [reply Providers/RustSyncManager.swift:56]: Noting something that needs to be fixed later, which is true for all of the other instances I've used it except here. I'm just going to remove XXX so it doesn't cause confusion.

File: Providers/RustSyncManager.swift [hunk 15]
```diff
@@ -0,0 +1,628 @@
...
+    ) -> Timer {
+        return Timer.scheduledTimer(timeInterval: interval,
+                                    target: self,
+                                    selector: selector,
+                                    userInfo: nil,
+                                    repeats: true)
+    }
+
+    func syncEverythingSoon() {
+        self.doInBackgroundAfter(SyncConstants.SyncOnForegroundAfterMillis) {
```
Comment (tarikeshaq) [inline Providers/RustSyncManager.swift:98]: if I'm understanding correctly, the backgrounding of this is purely for the purpose of the delay correct? Like the actual syncing will dispatch back to the `syncQueue` (which in turn dispatches back to the `Dispatch.global` default queue?

File: Providers/RustSyncManager.swift [hunk 16]
```diff
@@ -0,0 +1,628 @@
...
+        if !hasSynced && status == .authError {
+            return .warning(message: .FirefoxSyncOfflineTitle)
+        } else if !hasSynced && status == .networkError {
+            return .bad(message: .FirefoxSyncOfflineTitle)
+        } else {
+            return .good
+        }
+    }
+
+    fileprivate func endRustSyncing(_ result: MozillaAppServices.SyncResult) {
```
Comment (tarikeshaq) [inline Providers/RustSyncManager.swift:190]: naming nit: how about `endSyncing`? I see we use `startSyncing` etc

File: Providers/RustSyncManager.swift [hunk 17]
```diff
@@ -0,0 +1,628 @@
...
+        logger.log("Ending all syncs.",
+                   level: .info,
+                   category: .sync)
+
+        syncDisplayState = resolveSyncState(result: result)
+
+        #if MOZ_TARGET_CLIENT
```
Comment (tarikeshaq) [inline Providers/RustSyncManager.swift:197]: Do we need this and the `canSendUsageData`? I thought that with glean, telemetry settings are global (i.e glean won't send anything at all if disabled)?

File: Providers/RustSyncManager.swift [hunk 18]
```diff
@@ -0,0 +1,628 @@
...
+                    self.profile.places.registerWithSyncManager()
+                    registeredPlaces = true
+                }
+                rustEngines.append("bookmarks")
+            case "history":
+                if !registeredPlaces {
+                    self.profile.places.registerWithSyncManager()
+                    registeredPlaces = true
+                }
+                rustEngines.append("history")
```
Comment (tarikeshaq) [inline Providers/RustSyncManager.swift:396]: nit:
```suggestion
            case "bookmarks", "history":
                if !registeredPlaces {
                    self.profile.places.registerWithSyncManager()
                    registeredPlaces = true
                }
                rustEngines.append(engine)
```

File: Providers/RustSyncManager.swift [hunk 19]
```diff
@@ -0,0 +1,628 @@
...
+                           name: .ProfileDidFinishSyncing,
+                           object: nil)
+    }
+
+    func doInBackgroundAfter(_ millis: Int64, _ block: @escaping () -> Void) {
+        let queue = DispatchQueue.global(qos: DispatchQoS.background.qosClass)
+        // Pretty ambiguous here. I'm thinking .now was DispatchTime.now() and not
+        // Date.now()
+        queue.asyncAfter(
+            deadline: DispatchTime.now() + DispatchTimeInterval.milliseconds(Int(millis)),
```
[the remark this answers is not in the document]
Comment (lougeniaC64) [reply Providers/RustSyncManager.swift:251]: More code from `BrowserSyncManager`. This comment was [added 10 months ago](https://github.com/mozilla-mobile/firefox-ios/blame/a9e338a97059600a5f376e680a99c87c2d49abca/Providers/Profile.swift#L898) so I wasn't sure, if this was relevant for upcoming work. I'll just remove it.

File: Providers/RustSyncManager.swift (hunk 5 above)
Comment (lougeniaC64) [reply Providers/RustSyncManager.swift:243]: 1. So they're currently sent from `BrowserSyncManger` in Profile.swift. And the idea is that the `RustSyncManger` class that I'm adding in this PR will replace `BrowserSyncManager` during the Nimbus sync manager integration rollout. So there will be a lot of duplicated logic in both classes, but only one will ever be in use.
2. I suppose `RustSyncManager` doesn't need to listen to those notifications but they still need to be sent because there are a number of places in the Client code where those notifications are being observed. So, for instance, the `beginSyncing` function could call `onStartSyncing` but it would still need to emit the notification. This is more borrowed code so I don't have strong opinions here.

File: Providers/RustSyncManager.swift [hunk 20]
```diff
@@ -0,0 +1,628 @@
...
+import AuthenticationServices
+import Common
+
+// Extends NSObject so we can use timers.
+public class RustSyncManager: NSObject, SyncManager {
+    // We shouldn't live beyond our containing BrowserProfile, either in the main app
+    // or in an extension.
+    // But it's possible that we'll finish a side-effect sync after we've ditched the
+    // profile as a whole, so we hold on to our Prefs, potentially for a little while
+    // longer. This is safe as a strong reference, because there's no cycle.
```
[the remark this answers is not in the document]
Comment (lougeniaC64) [reply Providers/RustSyncManager.swift:19]: So this was defintely copied as is so I'd defer to @nbhasin2 on this one, but now that we've removed `unowned` is anything in the comment accurate? Because now we're holding on to prefs and profile which seems to negate the premise of the comment.

And to amplify Tarik's point about the implications, are we comfortable with the effects of removing `unowned` here?

File: Providers/RustSyncManager.swift (hunk 17 above)
Comment (lougeniaC64) [reply Providers/RustSyncManager.swift:197]: Bet! I'll remove this.

File: Providers/RustSyncManager.swift (hunk 15 above)
Comment (lougeniaC64) [reply Providers/RustSyncManager.swift:98]: I only intend to use either `syncQueue` or `Dispatch.global` depending on NIsh's preference.

File: Providers/RustSyncManager.swift (hunk 7 above)
Comment (lmarceau) [reply Providers/RustSyncManager.swift:327]: I had a look, and turns out yes using `"""` will make the log multiline in our logger too. This could make it harder to understand logs. I would suggest just using standard strings instead with `"`

My little test:
Screen Shot 2023-03-06 at 5 38 38 PM

Screen Shot 2023-03-06 at 5 37 34 PM

File: Providers/RustSyncManager.swift (hunk 20 above)
[the remark this answers is not in the document]
Comment (nbhasin2) [reply Providers/RustSyncManager.swift:19]: Is all you are trying to do is access `Profile`?

[the remark this answers is not in the document]
Comment (lougeniaC64) [reply Providers/RustSyncManager.swift:19]: Yeah I want some of the member functions and properties of the given profile.

File: Providers/RustSyncManager.swift (hunk 7 above)
Comment (lougeniaC64) [reply Providers/RustSyncManager.swift:327]: Sounds good! The only reason I used this format was to prevent the lines from being too long. 😊

File: Providers/RustSyncManager.swift (hunk 20 above)
[the remark this answers is not in the document]
Comment (nbhasin2) [reply Providers/RustSyncManager.swift:19]: When once you have that initialization ready in another PR we can double-check to make sure if this will have a strong reference cycle or not. A class alone won't have any issues.

File: Providers/RustSyncManager.swift (hunk 7 above)
Comment (lmarceau) [reply Providers/RustSyncManager.swift:327]: Yeah it's always a battle with strings to try to make them not too long haha. My only way otherwise to make them shorter is using variables like
```
let engines = "\(engineEnablements.compactMap { $0.value ? $0.key : nil })"
logger.log("engines to enable: \(engines)",
[...]
```

File: Providers/RustSyncManager.swift [hunk 21]
```diff
@@ -0,0 +1,628 @@
...
+
+    init(profile: BrowserProfile,
+         logger: Logger = DefaultLogger.shared) {
+        self.profile = profile
+        self.prefs = profile.prefs
+        self.logger = logger
+
+        super.init()
+
+        let center = NotificationCenter.default
```
[the remark this answers is not in the document]
Comment (lougeniaC64) [reply Providers/RustSyncManager.swift]: I actually removed this statement entirely when I removed the notification listeners as the `center` property was no longer needed.

[the remark this answers is not in the document]
Comment (lougeniaC64) [reply Providers/RustSyncManager.swift]: Sorry, I refactored that code and lost the thread on why we needed the protocol, I'll add it.

**Comment by lougeniaC64** [comment]
> Did another pass and dropped my thoughts going throughout 🙏 Most of it is questions and nits - I do think that testing this will be important
> 
> If I remember correctly, we wanted to get QA to test the branch before it lands on main, correct? Do we want to add any other automated tests, or do the ones covering SyncManager cover all the useful cases? (the sync what you want settings, all the meta request, clients syncing, etc

I've created a QA ticket for testing sync and send tab. I'm not sure exactly when in the process QA testing will happen, but I've been communicating with that team and they are expecting these changes. @nbhasin2 When can all of the changes in the sync manager epic branch be QA'd? Not sure what your plans are for merging it.

In regards to automated testing, the sync integration tests should still apply here. I will add some unit tests for the new logic I've added. Since the sync manager component handles client syncing on all syncs implicitly, I don't think that needs to be tested here.

File: Providers/RustSyncManager.swift (hunk 21 above)
[the remark this answers is not in the document]
Comment (lmarceau) [reply Providers/RustSyncManager.swift]: Perfect! As for the protocol, indeed there's not `object` param on the protocol. If we ever need it we can always add it. Thank you!

File: Providers/RustSyncManager.swift (hunk 12 above)
[the remark this answers is not in the document]
Comment (nbhasin2) [reply Providers/RustSyncManager.swift:378]: Ah, so we shouldn't create custom queues unless we need it. This is the mantra I go by and is what Apple agrees too. Ref: https://developer.apple.com/documentation/dispatch/dispatchqueue

However, there can be needs like 
- Performing task in a thread (syncQueue) that only your app is aware of
- Suspending the thread (syncQueue)
- Put a label so you can associate code specifically to your thread (syncQueue)

If you are doing any of this then we can leave `syncQueue` in our code. 

**IMPT Note:** Leaving it the way it is will not cause any harm but we should at least change `com.mozilla.firefox.sync` which is the same as Profile label to  `com.mozilla.firefox.sync.rust` to debug issues related to RustSyncManager 

**Final Thoughts** Change the label and leave the queue the way it is and we can come back to it for further review on how to leverage `syncQueue` and if we can't find anything in future then we use the Global Dispatch Queue

File: Providers/RustSyncManager.swift [hunk 22]
```diff
@@ -0,0 +1,574 @@
...
+            } else {
+                prefs.removeObjectForKey(PrefsKeys.KeyLastSyncFinishTime)
+            }
+        }
+    }
+
+    lazy var syncManagerAPI = RustSyncManagerAPI(logger: logger)
+
+    public func migrateSyncData() {
+        // The sync functions used by the old native sync manager did not return a
```
Comment (mhammond) [inline Providers/RustSyncManager.swift:48]: I think I probably mislead you here before but this comment isn't quite correct.

The persisted global state is just a [list of engines which have been declined](https://github.com/mozilla/application-services/blob/main/components/sync15/src/client/state.rs#L42-L62) - there's no per-engine state, so it will not cause anything to think there was a first-sync.

The main risk of not doing this here will be that the first "real" sync will supply [enabled_changes](https://github.com/mozilla/application-services/blob/main/components/sync_manager/src/types.rs#L14) if the user happens to have enabled/disabled an engine exactly as this upgrade was happening, and because we didn't accurately know what was "declined" before we might do the wrong thing. The fact we can be 100% sure this first sync doesn't supply that arg means we can't get confused.

However, there's actually a reasonable chance that not asking for *any* engines to sync might do the right thing here too - ie, if we can somehow tell this magic we want to sync `SyncEngineSelection::Some{ engines: vec![] }` I suspect it will do what we want.

However, ISTM there's still a bit of a risk here - what if this sync fails due to, say, a transient network error? (I can't see where it's actually called though!) We are then lumping edge-cases on edge-cases so the chances of users doing a wrong thing might be small enough we don't really care. Alternatively, we could almost certainly construct a valid `persistedState` from what we know the enabled/disabled state is already (it's just a json string), which is info we already have - then we could avoid this entirely. (But is even *that* true - for all we know, the user tried to change the enabled state just before we upgraded, but that sync failed - so we can't be 100% sure the state of the UI checkboxes are an accurate representation of the current server state can we?

I wonder if another option is to simply not pass that `enabled_changes` param on a first sync based on a pref or something? It looks like that would mean this can die entirely too, and will still do the right thing after that first sync? I think that might work except for the caveat about that sync failing?

tl;dr - this needs more thought. Maybe we should start by working out exactly what might happen if that edge-case of the user changing a checkbox before the first normal sync happens and whether we are fine with that edge-case given it seems quite unlikely

File: Shared/Prefs.swift [hunk 23]
```diff
@@ -9,6 +9,13 @@ public struct PrefsKeys {
     // When this pref is set (by the user) it overrides default behaviour which is just based on app locale.
     public static let KeyEnableChinaSyncService = "useChinaSyncService"
     public static let KeyLastRemoteTabSyncTime = "lastRemoteTabSyncTime"
+
+    // Global sync state for rust sync manager
+    public static let RustSyncManagerPersistedState = "rustSyncManagerPersistedStatKey"
```
Comment (mhammond) [inline Shared/Prefs.swift:14]: Is `StatKey` a typo?

File: Shared/Prefs.swift [hunk 24]
```diff
@@ -9,6 +9,13 @@ public struct PrefsKeys {
...
+
+    // Persisted State for rust sync manager has been migrated
+    public static let MigratedRustSyncPersistedState = "migratedRustSyncPersistedStateKey"
```
Comment (mhammond) [inline Shared/Prefs.swift:17]: what's this one used for?

**Review by mhammond — COMMENTED** [review]
LGTM, just still a bit confused about that migration stuff which we need to think through a bit more.

File: Shared/Prefs.swift (hunk 24 above)
Comment (lougeniaC64) [reply Shared/Prefs.swift:17]: It's a flag for whether `migrateSyncData` has been  run. It's now obsolete since I'm removing that function, but basically `migrateSyncData` wouldn't be executed if  `MigratedRustSyncPersistedState` was true.

File: Providers/RustSyncManager.swift (hunk 22 above)
Comment (lougeniaC64) [reply Providers/RustSyncManager.swift:48]: Thanks for the clarification. So stepping back a bit, and I might be missing something here, but as the comment suggests I thought the reason for performing this "migration" was to prevent the first sync with the sync manager component taking longer than normal. If the actual issue is the edge case of the user changing their local engine enablements between the time we retrieve them and when they're passed into the sync call, is this an issue that's specific to the sync manager integration? In other words, is this an issue that could happen now in Firefox iOS release (or any other platform)?

I think for now I'm going to remove this function so that resolving how we handle this situation doesn't block the whole effort. In the meantime, I'll see if I can replicate this locally. Let me know if you have any objections @mhammond.

**Comment by nbhasin2** [comment]
@lougeniaC64 I can go in circles to improve the old code however as we discussed offline let's make the following changes for this PR 

a) make sure profile is marked `weak` and we can try to see if that causes any issues. Ideally, we wouldn't mark it `weak` but replace it with a delegate. Something like below but for now lets go with `weak` and come back to it to have the changes so we don't even have profile. This is to avoid `RustSyncManager` having direct access to `profile`
image

b) Update queue names as those should be different for better debugging and if you don't need it you can simply use the global dispatch queue. 

Note: I believe you have already made these changes so I will one last walkthrough of this PR once the conflicts are resolved

File: Providers/RustSyncManager.swift [hunk 25]
```diff
@@ -0,0 +1,553 @@
...
+                        self.syncManagerAPI.sync(params: params) { syncResult in
+                            guard syncResult.status == ServiceStatus.ok else {
+                                deferred.fill(Maybe(success: syncResult))
+                                self.endSyncing(syncResult)
+                                return
+                            }
+
+                            // Save the persisted state
+                            self.prefs
+                                .setString(syncResult.persistedState,
```
Comment (mhammond) [inline Providers/RustSyncManager.swift:411]: It's a bit of a bug in the rust code - `syncResult.persistedState` should probably be nullable to make the intent clearer, but I think you should persist this value any time it is not empty, even if the sync "failed"

File: Providers/RustSyncManager.swift [hunk 26]
```diff
@@ -0,0 +1,553 @@
...
+                            if let declined = syncResult.declined {
+                                RustTogglableEngines.forEach({
+                                    if declined.contains($0) {
+                                        updateEnginePref($0, false)
+                                    } else {
+                                        updateEnginePref($0, true)
+                                    }
+                                })
+                            } else {
+                                RustTogglableEngines.forEach({
```
Comment (mhammond) [inline Providers/RustSyncManager.swift:448]: This branch should be removed - it is nullable, and `null` will just mean we failed before we could determine what the error is. The branch above (ie, where you iterate over `declined` if it is non-null) should be done even if the success value isn't ok.

File: Providers/RustSyncManager.swift [hunk 27]
```diff
@@ -0,0 +1,553 @@
...
+                                return
+                            }
+
+                            // Save the persisted state
+                            self.prefs
+                                .setString(syncResult.persistedState,
+                                           forKey: PrefsKeys.RustSyncManagerPersistedState)
+
+                            let declinedEngines = String(describing: syncResult.declined ?? [])
+                            let telemetryData = syncResult.telemetryJson ??
```
Comment (mhammond) [inline Providers/RustSyncManager.swift:415]: ditto here - we want to report failures too.

**Review by mhammond — APPROVED** [review]
Thanks - this looks good although I think it needs a few minor tweaks for the sync failing scenario.

I'll try and see if my iOS build errors magically went away and give it a test, but if you can verify the declined engine stuff appears to work ok for you then I'm fine with landing it if you are 🚀

File: Providers/RustSyncManager.swift (hunk 27 above)
Comment (lougeniaC64) [reply Providers/RustSyncManager.swift:415]: That `telemetryData` variable is for the logging. The telemetry is reported in the `endSyncing` function. Worth noting that the temp telemetry doesn't report status, which is just another reason for it to go.

File: Providers/RustSyncManager.swift [hunk 28]
```diff
@@ -0,0 +1,552 @@
...
+
+    @objc func syncOnTimer() {
+        syncEverything(why: .scheduled)
+        profile?.pollCommands()
+    }
+
+    fileprivate func repeatingTimerAtInterval(
+        _ interval: TimeInterval,
+        selector: Selector
+    ) -> Timer {
```
Comment (nbhasin2) [inline Providers/RustSyncManager.swift:61]:
```suggestion
    private func repeatingTimerAtInterval(_ interval: TimeInterval,
                                              selector: Selector) -> Timer {
```

File: Providers/RustSyncManager.swift [hunk 29]
```diff
@@ -0,0 +1,552 @@
...
+        if since > SyncConstants.SyncOnForegroundMinimumDelayMillis {
+            syncEverythingSoon()
+        }
+    }
+
+    public func applicationDidEnterBackground() {
+        backgrounded = true
+    }
+
+    fileprivate func beginSyncing() {
```
Comment (nbhasin2) [inline Providers/RustSyncManager.swift:138]: nit: swift update
```suggestion
    private func beginSyncing() {
```

File: Providers/RustSyncManager.swift [hunk 30]
```diff
@@ -0,0 +1,552 @@
...
+        syncDisplayState = .inProgress
+        notifySyncing(notification: .ProfileDidStartSyncing)
+    }
+
+    private func resolveSyncState(
+        result: MozillaAppServices.SyncResult
+    ) -> SyncDisplayState {
```
Comment (nbhasin2) [inline Providers/RustSyncManager.swift:145]: nit: indent
```suggestion
    private func resolveSyncState(result: MozillaAppServices.SyncResult) -> SyncDisplayState {
```

File: Providers/RustSyncManager.swift [hunk 31]
```diff
@@ -0,0 +1,552 @@
...
+        if !hasSynced && status == .authError {
+            return .warning(message: .FirefoxSyncOfflineTitle)
+        } else if !hasSynced && status == .networkError {
+            return .bad(message: .FirefoxSyncOfflineTitle)
+        } else {
+            return .good
+        }
+    }
+
+    fileprivate func endSyncing(_ result: MozillaAppServices.SyncResult) {
```
Comment (nbhasin2) [inline Providers/RustSyncManager.swift:163]: nit: swift update
```suggestion
    private func endSyncing(_ result: MozillaAppServices.SyncResult) {
```

File: Providers/RustSyncManager.swift [hunk 32]
```diff
@@ -0,0 +1,552 @@
...
+
+    func canSendUsageData() -> Bool {
+        return profile?.prefs.boolForKey(AppConstants.prefSendUsageData) ?? true
+    }
+
+    private func notifySyncing(notification: Notification.Name) {
+        notificationCenter.post(name: notification)
+    }
+
+    init(profile: BrowserProfile,
```
Comment (nbhasin2) [inline Providers/RustSyncManager.swift:198]: nit: usually inits are at the beginning but in this file its way in the end, would be nice if you can move it more up

File: Providers/RustSyncManager.swift [hunk 33]
```diff
@@ -0,0 +1,552 @@
...
+    }
+
+    func doInBackgroundAfter(_ millis: Int64, _ block: @escaping () -> Void) {
+        let queue = DispatchQueue.global(qos: DispatchQoS.background.qosClass)
+        queue.asyncAfter(
+            deadline: DispatchTime.now() + DispatchTimeInterval.milliseconds(Int(millis)),
+            execute: block)
+    }
+
+    var prefsForSync: Prefs {
```
Comment (nbhasin2) [inline Providers/RustSyncManager.swift:216]: nit: could we move this var near the top where other variables are initialized

File: Providers/RustSyncManager.swift [hunk 34]
```diff
@@ -0,0 +1,552 @@
...
+        public let description = "Failed to get token server endpoint url."
+    }
+
+    public class EngineAndKeyRetrievalError: MaybeErrorType {
+        public let description = "Failed to get sync engine and key data."
+    }
+
+    fileprivate func getEnginesAndKeys(
+        engines: [String]
+    ) -> Deferred<Maybe<([EngineIdentifier], [String: String])>> {
```
Comment (nbhasin2) [inline Providers/RustSyncManager.swift:309]:
```suggestion
    private func getEnginesAndKeys(engines: [String]) -> Deferred<Maybe<([EngineIdentifier],
                                                                             [String: String])>> {
```

File: Providers/RustSyncManager.swift [hunk 35]
```diff
@@ -0,0 +1,552 @@
...
+        }
+
+        deferred.fill(Maybe(success: (rustEngines, localEncryptionKeys)))
+        return deferred
+    }
+
+    fileprivate func syncRustEngines(
+        why: MozillaAppServices.SyncReason,
+        engines: [String]
+    ) -> Deferred<Maybe<MozillaAppServices.SyncResult>> {
```
Comment (nbhasin2) [inline Providers/RustSyncManager.swift:348]: nit: indent 

```suggestion
    private func syncRustEngines(why: MozillaAppServices.SyncReason,
                                 engines: [String]) -> Deferred<Maybe<MozillaAppServices.SyncResult>> {
```

File: Providers/RustSyncManager.swift [hunk 36]
```diff
@@ -0,0 +1,552 @@
...
+                    }
+                }
+            }
+        }
+        return deferred
+    }
+
+    fileprivate func toSyncManagerDeviceType(
+        deviceType: DeviceType
+    ) -> SyncManagerDeviceType {
```
Comment (nbhasin2) [inline Providers/RustSyncManager.swift:463]:
```suggestion
    private func toSyncManagerDeviceType(deviceType: DeviceType) -> SyncManagerDeviceType {
```

**Comment by nbhasin2** [comment]
@lougeniaC64 I see that at times you are using `MozillaAppServices.SyncResult` which could make the line longer. Below is a small suggestion to add it top the top of file and use it below. Let me know what you think?

image

typealias: `private typealias MZSyncResult = MozillaAppServices.SyncResult`

Before
```
private func resolveSyncState(result: MozillaAppServices.SyncResult) -> SyncDisplayState {
```

After
```
private func resolveSyncState(result: MZSyncResult) -> SyncDisplayState {
```

**Review by nbhasin2 — APPROVED** [review]
Giving a preliminary approval as others are nits but we should revisit this in future to take a look at other areas we can improve. 

@lougeniaC64 Thanks for addressing so many comments, you are a trooper 🔥
:shipit:

**Review by nbhasin2 — APPROVED** [review]

**Comment by nbhasin2** [comment]
@lougeniaC64 thanks for creating the ticket, I am going to merge this into the epic branch
